### PR TITLE
Add AP_Datalogger: a library for logging none flight sensors

### PR DESCRIPTION
**Pull request by IamPete1**
This work was commissioned by ISS Aerospace.

This adds a new library for data logging. This is to be used for logging data that is not needed for operation and is not needed in real time. 

The first back end is a serial data logger for logging ASCII data. The data received should be terminated by a \r or \n. 

This has been tested on a custom sensor but can also be used to log NMEA data (tested) for example.

see [readme](https://github.com/IamPete1/ardupilot/blob/CSV_log/libraries/AP_DataLogger/README.md)

This is my first go at #if including a lib. It is not included for plane or copter if minimize feature is true, it is always included in rover. I have omitted it from sub and tracker.

### CODE REVIEW

File: libraries/AP_DataLogger/AP_DataLogger.cpp [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+    if (_enabled == 0) {
+        clear();
+        return;
+    }
+
+    if (_logger == nullptr) {
+        clear();
+        return;
+    }
+    if (!_logger->should_log(4294967295)) { // this is all ones 32 bit bitmask, ie allways log unless disabled
```
Comment (IamPete1) [inline libraries/AP_DataLogger/AP_DataLogger.cpp]: This is a odd way of seeing if we should log but it was the best I could think of. Seems redundant to have a bitmask bit, you might as well just turn off the lib. I just wanted to know if logging of any sort was active.

Comment (WickedShell) [reply libraries/AP_DataLogger/AP_DataLogger.cpp]: `0xFFFFFFFF` is probably a lot clearer :)

**Comment by rmackay9** [comment]
I've also had a couple of people ask me (even quite recently) whether they could log some extra data to the dataflash.

**Comment by IamPete1** [comment]
@WickedShell I hadn't had a scripting master class when I wrote this ;)

I also have plans to add some less generic back-ends for i2c sensors. I think that might be harder via scripting? (I have never done a i2c driver so it must be hard right?)

File: libraries/AP_DataLogger/AP_DataLogger.cpp (hunk 1 above)
Comment (IamPete1) [reply libraries/AP_DataLogger/AP_DataLogger.cpp]: changed decimal to hex pending any nicer way than should_log()

**Comment by IamPete1** [comment]
rebased

**Comment by tridge** [comment]
on dev call we've decided this is much better done as a script
@WickedShell  will help with serial abstraction

**Comment by WickedShell** [comment]
@IamPete1 given that we have the scripting stuff in (and you have part of this managed in #13268) can this one be closed?

**Comment by rmackay9** [comment]
My guess is that functionally what we want now is to be able to log from a script.. but that's probably quite a different thing than what's in this PR

**Comment by WickedShell** [comment]
@rmackay9 scripts can't log to the dataflash log, but they are actually able to read/write to files, so sensor data can be logged to it's own file easily. (It can even mimic the DF format if it really wants). It's not quite the full equivalent to what's here, but NMEA information in  a DF log will need extra processing to be useful already so I don't think it'd be painful.

**Comment by IamPete1** [comment]
@rmackay9 I plan to do a example script that duplicates this functionality then I'm happy to close.
@WickedShell I was going to add a binding to get stuff into the data flash log so its all in one place. Logging to a file would be nice too though, are then any examples / instructions on how to do that?

**Comment by WickedShell** [comment]
@IamPete1 There aren't any in repo examples, it's the io calls [here](https://www.lua.org/manual/5.3/manual.html#6.8). Generally any tutorial you find on lua file stuff will be correct.

**Comment by IamPete1** [comment]
closed by https://github.com/ArduPilot/ardupilot/pull/13315

**Pull request closed without merging**